### PR TITLE
fix: Add migration to add created_by_fk as explicit owner for charts and datasets

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

https://github.com/apache/superset/pull/19854 introduced frontend logic which would check that (for Alpha users) only owners can edit datasets. The issue is the definition of ownership between the frontend and backend differs. Specifically the frontend merely checks whether the user is listed as a owner—per the [owners](https://github.com/apache/superset/blob/master/superset/connectors/sqla/models.py#L694) relationship—whereas the backend uses the [check_ownership](https://github.com/apache/superset/blob/3483446c28113fb9eb1e8adb8978e305ff5e7c65/superset/views/base.py#L647-L691) method which checks (in addition to the `owner` and `owners` relationships) whether the user is the creator. This dichotomy means that creators who are not explicitly listed as owners are unable to edit their datasets.

Unwinding the logic surfaces additional observations regarding ownership for various assets, i.e., charts, dashboards, datasets, and reports:

1. The DAO logic correctly adds the creator as an explicit owner ([example](https://github.com/apache/superset/blob/master/superset/datasets/commands/create.py#L92-L93)) so this isn't an issue for any asset which was created via a DAO commend.
2. Prior to the DAO logic only dashboards contained [pre_add](https://github.com/apache/superset/blob/8e29ec5a6685867ffc035d20999c54c2abe36fb1/superset/views/dashboard/views.py#L96-L97) logic which added the creator as an explicit owner meaning that this issue only impacts historical charts and datasets.
3. The `check_ownership` method contained a check for an `owner` relationship however grepping the code per `git grep "\bowner = relationship\b"` yielded no matches, i.e., it served no purpose.

Given these insights this PR performs the following:

1. It adds a database migration to add all creators as owners (if not already listed) for both charts and datasets. Granted this may be a little presumptive, i.e., we may be re-adding someone as an "owner" who was previously removed, however there's no way to determine what the intent of the current database state is.
2. It updates the `check_ownership` method to remove checking the `owner` (obsolete) and `created_by` (addressed in the migration) fields.

Finally a few things worth noting:

1. The frontend likely should rely on an API call to check ownership to adhere to the DRY principle and ensure that the frontend and backend logic for ownership is consistent.
2. The backend SQLAlchemy `pre_add`, `pre_update`, etc. checks should likely be deprecated given the logic also resides in the DAO commands. This really muddies the code logic and makes it hard to grok how things work (or should work).
3. I'm not really sure why the DAO `validate` command also populates fields, i.e., adds owners. This seems somewhat contradictory and non-intuitive, i.e., validation shouldn't update the model. I gather it was likely for efficiency reasons, but I sense there is merit in rewriting the DAO logic to break up the validation and population phases.

### TESTING INSTRUCTIONS

Tested the migration and confirmed that, 

```sql
SELECT 
  s.id, 
  s.created_by_fk
FROM 
  slices s
LEFT OUTER JOIN 
  slice_user su
ON 
 s.id = su.slice_id AND 
 s.created_by_fk = su.user_id  
WHERE 
  su.slice_id IS NULL
```

,

```sql  
SELECT 
  sd.id, 
  sd.created_by_fk
FROM 
   sl_datasets sd
LEFT OUTER JOIN 
  sl_dataset_users sdu
ON 
 sd.id = sdu.dataset_id AND 
 sd.created_by_fk = sdu.user_id  
WHERE 
  sdu.dataset_id IS NULL
```

and

```sql  
SELECT 
  t.id, 
  t.created_by_fk
FROM 
  tables t
LEFT OUTER JOIN 
  sqlatable_user su
ON 
 t.id = su.table_id AND 
 t.created_by_fk = su.user_id  
WHERE 
  su.table_id IS NULL
```

returned no rows.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/migrations/versions/2022-07-05_15-48_409c7b420ab0_add_created_by_fk_as_owner.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            .outerjoin(
+                SqlaTableUser,
+                and_(
+                    SqlaTableUser.table_id == Table.id,
+                    SqlaTableUser.user_id == Table.created_by_fk,
+                ),
+            )
+            .filter(SqlaTableUser.table_id == None),
+        )
+    )
```
Comment (ktmud) [inline superset/migrations/versions/2022-07-05_15-48_409c7b420ab0_add_created_by_fk_as_owner.py:100]: Should we do the same for the new dataset models, too? As the shadow-writing is already in process.

https://github.com/apache/superset/blob/231716cb50983b04178602b86c846b7673f9d8c3/superset/datasets/models.py#L65-L70

**Review by ktmud — COMMENTED** [review]
The change makes sense to me.

**Comment by ktmud** [comment]
> we may be re-adding someone as an "owner" who was previously removed

Since the problem seems to be a legacy dataset not created by DAO now has an empty owners list because the creator was not explicitly added, I wonder if it is safer to just set the creator as the sole owner when the owners list is empty?

**Comment by john-bodley** [comment]
@ktmud per your comment,

> Since the problem seems to be a legacy dataset not created by DAO now has an empty owners list because the creator was not explicitly added, I wonder if it is safer to just set the creator as the sole owner for datasets/slices where the owners list is empty?

I hear what you're saying, though the old `check_ownership` logic was based on always including the creator as an owner regardless of who was listed and thus I think you could argue either way which approach is best—acknowledging that neither are perfect—, i.e., the current migration ensures code logic parity whereas your suggestion likely reduces the level of churn in terms of re-adding the creator as an owner who previously was removed.

File: superset/migrations/versions/2022-07-05_15-48_409c7b420ab0_add_created_by_fk_as_owner.py (hunk 1 above)
Comment (john-bodley) [reply superset/migrations/versions/2022-07-05_15-48_409c7b420ab0_add_created_by_fk_as_owner.py:130]: Thanks @ktmud for the tip. I've updated the migration and re-tested.

**Comment by john-bodley** [comment]
ping @ktmud

**Comment by ktmud** [comment]
Do we need to a rebase? There is recently a new migration merged.

**Review by ktmud — APPROVED** [review]

**Comment by AAfghahi** [comment]
Hello, just to flag this. When I tried to do a migration with my database that has just examples, this required me to drop the database and then re-populate.

**Comment by ktmud** [comment]
@AAfghahi can you share the error messages you got?

**Comment by AAfghahi** [comment]
Hey, sorry I thought that I wrote in here but forgot to press comment apparently. 

I did not save the error message unfortunately, however. I remember that it would hit this migration and then error out because `sl_dataset_user` (which I think is the new fk) was null and that caused an error. The sql alchemy error said that it was most likely a database error. 
``

**Pull request merged**